### PR TITLE
DUPLO-13194 Support for prepended user data on hosts

### PR DESCRIPTION
**Pull request by duplodavid**
## Change description

- Adds a `prepend_user_data` attribute to the `duplocloud_aws_host` and `duplocloud_asg_profile` resources.
- This utilizes the `IsCombinedUserData` boolean property in the backend APIs of the same meaning.
- The data source types for these resources also have the new attribute added to them.
- Avoids triggering an unnecessary "force replacement" on hosts and ASGs that prepend Duplo's user data since the actual user data on the instance will always differ from the provided custom user data. This is the same behavior of `base64_user_data` prior to adding this new attribute.
- "Force replacement" still occurs when the custom user data in the `base64_user_data` attribute is modified in the TF.
- The default for `prepend_user_data` is `false`.
- Added the `customdiff` helper used in the implementation which was missing by running `go mod vendor`.

There is a backend issue I discovered which causes the `IsCombinedUserData` to come back as `false` for hosts even if `true` was sent and Duplo's user data did get prepended. While that is being fixed, it is safe to ignore the value that comes back on the response. It only matters during creation and would trigger a "force replacement" if changed.

## Type of change
- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)

## Related issues

Fix [DUPLO-13194](https://app.clickup.com/t/8655600/DUPLO-13194)

___

## **Type**
enhancement

___

## **Description**
- Introduced `prepend_user_data` attribute to `duplocloud_aws_host` and `duplocloud_asg_profile` resources for prepending user data on hosts.
- Added custom diff logic to avoid unnecessary "force replacement" for hosts and ASGs with prepended Duplo's user data.
- Updated data source types for `duplocloud_aws_host` and `duplocloud_asg_profile` to include the new `prepend_user_data` attribute.
- Enhanced documentation to reflect the addition of the `prepend_user_data` attribute.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>data_source_duplo_asg_profile.go</strong><dd><code>Add Prepend User Data Attribute to ASG Profile Data Source</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplocloud/data_source_duplo_asg_profile.go

- Added `prepend_user_data` attribute to the data source schema.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/393/files#diff-3338765c0215fecbec2e7b62b9d3f83e83887af6db36c74d7e38d3b80670d8d6">+6/-6</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>data_source_duplo_native_hosts.go</strong><dd><code>Add Prepend User Data Attribute to Native Hosts Data Source</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplocloud/data_source_duplo_native_hosts.go

- Added `prepend_user_data` attribute to the data source schema.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/393/files#diff-380c7b3c31eee66e687a04f041c3c64d3f2454fda0f90a2d06eff379ba15b061">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>resource_duplo_asg_profile.go</strong><dd><code>Enhance ASG Profile Resource with Prepend User Data Support</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplocloud/resource_duplo_asg_profile.go

<li>Added <code>prepend_user_data</code> attribute to the resource schema.<br> <li> Introduced custom diff logic for user data.<br> <li> Added initialization for user data options.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/393/files#diff-f58c7134634373bfbd4607fa72e0d5fb28a0da60d136394bacd5671ea75c8346">+10/-2</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>resource_duplo_aws_host.go</strong><dd><code>Enhance AWS Host Resource with Prepend User Data Support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplocloud/resource_duplo_aws_host.go

<li>Added <code>prepend_user_data</code> and <code>initial_base64_user_data</code> attributes to the <br>resource schema.<br> <li> Implemented custom diff logic for user data.<br> <li> Added initialization for user data options.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/393/files#diff-c6eb5a4b1c897b209d4d26aeb30aefbd2a7560e52fe9040378f94b9a7bf057b3">+47/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>asg.go</strong><dd><code>SDK Support for Prepend User Data in ASG Profile</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplosdk/asg.go

- Added `PrependUserData` field to `DuploAsgProfile` struct.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/393/files#diff-b9e34e160487d98157b8b887377449bc8b90d93a1d7640c770a57ab0a7ca5507">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>native_hosts.go</strong><dd><code>SDK Support for Prepend User Data in Native Host</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
duplosdk/native_hosts.go

- Added `PrependUserData` field to `DuploNativeHost` struct.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/393/files#diff-0edc1b188d4b47dcfe9e47c244e961a3f1e85b0bf37d0f740d167aa654951e76">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Update Changelog with Prepend User Data Feature</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
CHANGELOG.md

<li>Documented new <code>prepend_user_data</code> attribute and related enhancements.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/393/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>asg_profiles.md</strong><dd><code>Document Prepend User Data Attribute in ASG Profile Data Source</code></dd></summary>
<hr>
      
docs/data-sources/asg_profiles.md

<li>Documented <code>prepend_user_data</code> attribute in ASG profile data source.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/393/files#diff-52b39cccffadb88ed9ceb2dfa500b366c11f840a94fb3ef00945abf6875916b0">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>native_hosts.md</strong><dd><code>Document Prepend User Data Attribute in Native Hosts Data Source</code></dd></summary>
<hr>
      
docs/data-sources/native_hosts.md

<li>Documented <code>prepend_user_data</code> attribute in native hosts data source.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/393/files#diff-f3666853267805ff4889c0f5278a6b065703bcb823563033c5a4cd04d0ac2099">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>asg_profile.md</strong><dd><code>Document Prepend User Data Attribute in ASG Profile Resource</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
docs/resources/asg_profile.md

- Documented `prepend_user_data` attribute in ASG profile resource.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/393/files#diff-0b0507a0a18d1275d81b42d11e06ba436e7ef85ec5c0cb56cc83b8e6381c2909">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>aws_host.md</strong><dd><code>Document Prepend User Data Attribute in AWS Host Resource</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
docs/resources/aws_host.md

- Documented `prepend_user_data` attribute in AWS host resource.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/393/files#diff-61d13b5b9550924822c5ddb133bfddb318d2a60477baadda034f9994c9492915">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Inline File Walkthrough 💎</strong></summary><hr>

For enhanced user experience, the `describe` tool can add file summaries directly to the "Files changed" tab in the PR page.
This will enable you to quickly understand the changes in each file, while reviewing the code changes (diffs).

To enable inline file summary, set `pr_description.inline_file_summary` in the configuration file, possible values are:
- `'table'`: File changes walkthrough table will be displayed on the top of the "Files changed" tab, in addition to the "Conversation" tab.
- `true`: A collapsable file comment with changes title and a changes summary for each file in the PR.
- `false` (default): File changes walkthrough will be added only to the "Conversation" tab.
<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

**Comment by zafarabbas** [comment]
Task linked: [DUPLO-13194 TF: Add ability to prepend user data to host/ASG](https://app.clickup.com/t/8655600/DUPLO-13194)

File: duplocloud/resource_duplo_aws_host.go [hunk 1]
```diff
@@ -351,6 +383,18 @@ func resourceAwsHostCreate(ctx context.Context, d *schema.ResourceData, m interf
...
 }
 
+func initUserDataOptions(d *schema.ResourceData) {
+	// Store initial base64_user_data if supplied
+	if userData, ok := d.GetOk("base64_user_data"); ok {
+		d.Set("initial_base64_user_data", userData)
+	}
+
+	// Set default prepend_user_data value here to avoid a force replacement due to DUPLO-14926
+	if _, ok := d.GetOk("prepend_user_data"); !ok {
```
Comment (tahir-duplo) [inline duplocloud/resource_duplo_aws_host.go:393]: Can we set the default true to a schema level instead of here?

Comment (joek-duplo) [reply duplocloud/resource_duplo_aws_host.go:393]: I think the default is supposed to be `false`, actually.

**Review by joek-duplo — CHANGES_REQUESTED** [review]

File: duplocloud/resource_duplo_aws_host.go (hunk 1 above)
Comment (duplodavid) [reply duplocloud/resource_duplo_aws_host.go:393]: I've made it default to false and put it at the schema level. I hadn't put it there as a different kind of workaround to the read response for hosts always being false. Got rid of this logic too I'm glad to say. Now we can fix the minor backend issue later and it won't affect the behavior here.

**Review by joek-duplo — DISMISSED** [review]

**Pull request merged**